gobblin-data-management/src/main/java/org/apache/gobblin/data/management/copy/splitter/DistcpFileSplitter.java

```diff
@@ -0,0 +1,227 @@
+      return Lists.newArrayList(workUnit);
+    }
+
+    Collection<WorkUnit> newWorkUnits = Lists.newArrayList();
+
+    long lengthPerSplit = (maxSplitSize / blockSize) * blockSize;
+    int splits = (int) (len / lengthPerSplit + 1);
+
+    for (int i = 0; i < splits; i++) {
+      WorkUnit newWorkUnit = WorkUnit.copyOf(workUnit);
+
+      long lowPos = lengthPerSplit * i;
+      long highPos = lengthPerSplit * (i + 1);
+
+      Split split = new Split(lowPos, highPos, i, splits,
+          String.format("%s.__PART%d__", file.getDestination().getName(), i));
+      String serializedSplit = GSON.toJson(split);
+
+      newWorkUnit.setProp(SPLIT_KEY, serializedSplit);
+
+      Guid oldGuid = CopySource.getWorkUnitGuid(newWorkUnit).get();
+      Guid newGuid = oldGuid.append(Guid.fromStrings(serializedSplit));
+
+      CopySource.setWorkUnitGuid(workUnit, newGuid);
+      newWorkUnits.add(newWorkUnit);
+    }
+    return newWorkUnits;
+  }
+
+  /**
+   * Finds all split work units in the input collection and merges the file parts into the expected output files.
+   * @param fs {@link FileSystem} where file parts exist.
+   * @param workUnits Collection of {@link WorkUnitState}s possibly containing split work units.
+   * @return The collection of {@link WorkUnitState}s where split work units for each file have been merged.
+   * @throws IOException
+   */
+  public static Collection<WorkUnitState> mergeAllSplitWorkUnits(FileSystem fs, Collection<WorkUnitState> workUnits)
+      throws IOException {
+    ListMultimap<CopyableFile, WorkUnitState> splitWorkUnitsMap = ArrayListMultimap.create();
+    for (WorkUnitState workUnit : workUnits) {
+      if (isSplitWorkUnit(workUnit)) {
+        CopyableFile copyableFile = (CopyableFile) CopySource.deserializeCopyEntity(workUnit);
+        splitWorkUnitsMap.put(copyableFile, workUnit);
+      }
+    }
+
+    for (CopyableFile file : splitWorkUnitsMap.keySet()) {
+      log.info(String.format("Merging split file %s.", file.getDestination()));
+
+      WorkUnitState oldWorkUnit = splitWorkUnitsMap.get(file).get(0);
+      Path outputDir = FileAwareInputStreamDataWriter.getOutputDir(oldWorkUnit);
+      CopyEntity.DatasetAndPartition datasetAndPartition =
+          file.getDatasetAndPartition(CopySource.deserializeCopyableDataset(oldWorkUnit));
+      Path parentPath = FileAwareInputStreamDataWriter.getOutputFilePath(file, outputDir, datasetAndPartition)
+          .getParent();
+
+      WorkUnitState newWorkUnit = mergeSplits(fs, file, splitWorkUnitsMap.get(file), parentPath);
+
+      for (WorkUnitState wu : splitWorkUnitsMap.get(file)) {
+        // Set to committed so that task states will not fail
+        wu.setWorkingState(WorkUnitState.WorkingState.COMMITTED);
+        workUnits.remove(wu);
+      }
+      workUnits.add(newWorkUnit);
+    }
+    return workUnits;
+  }
+
+  /**
+   * Merges all the splits for a given file.
```

Comment:
Edit javadoc to add that this is called on the targetFs (after file data has been copied over to target fs).